Android/src/org/droidplanner/android/maps/providers/baidu_map/BaiduMapOfflineMapPreference.java

```diff
@@ -0,0 +1,357 @@
+            mLocalMapList = new ArrayList<MKOLUpdateElement>();
+        }
+
+        ListView localMapListView = (ListView) view.findViewById(R.id.localmaplist);
+        mLAdapter = new LocalMapAdapter();
+        localMapListView.setAdapter(mLAdapter);
+
+        Button button = (Button) view.findViewById(R.id.search);
+        button.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                search(v);
+            }
+        });
+
+        button = (Button) view.findViewById(R.id.start);
+        button.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                start(v);
+            }
+        });
+
+        button = (Button) view.findViewById(R.id.stop);
+        button.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                stop(v);
+            }
+        });
+
+        button = (Button) view.findViewById(R.id.del);
+        button.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                remove(v);
+            }
+        });
+
+        button = (Button) view.findViewById(R.id.clButton);
+        button.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                clickCityListButton(v);
+            }
+        });
+
+        button = (Button) view.findViewById(R.id.localButton);
+        button.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                clickLocalMapListButton(v);
+            }
+        });
+    }
+
+    /**
+     * switch to city list
+     *
+     * @param view
+     */
+    public void clickCityListButton(View view) {
+        if (mBindDialogView != null) {
+            mLocalMapLayout.setVisibility(View.GONE);
+            mCityListLayout.setVisibility(View.VISIBLE);
+        }
+    }
+
+    /**
+     * switch to download manager list
+     *
+     * @param view
+     */
+    public void clickLocalMapListButton(View view) {
+        if (mBindDialogView != null) {
+            mLocalMapLayout.setVisibility(View.VISIBLE);
+            mCityListLayout.setVisibility(View.GONE);
+        }
+    }
+
+    /**
+     * search offline city
+     *
+     * @param view
+     */
+    public void search(View view) {
+        ArrayList<MKOLSearchRecord> records = mOffline.searchCity(mCityNameView
+                .getText().toString());
+        if (records == null || records.size() != 1) {
+            return;
+        }
+        mCidView.setText(String.valueOf(records.get(0).cityID));
+    }
+
+    /**
+     * Start to download
+     *
+     * @param view
+     */
+    public void start(View view) {
+        int cityid = Integer.parseInt(mCidView.getText().toString());
+        mOffline.start(cityid);
```

Comment:
@droneboost is `mOffline.start(cityId)` a long running operation? Could it block the main thread?